File: backend/app/models/course.py
from sqlalchemy import Column, String, Float, DateTime, ForeignKey, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..database import Base

class Course(Base):
    __tablename__ = "courses"
    
    id = Column(String(36), primary_key=True)
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String(500), nullable=False)
    description = Column(Text)
    overall_progress = Column(Float, default=0.0)
    average_grade = Column(Float)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    documents = relationship("Document", back_populates="course")
    enrollments = relationship("Enrollment", back_populates="course")
    statistics = relationship("CourseStatistics", back_populates="course", uselist=False)
    activity_logs = relationship("ActivityLog", back_populates="course")

File: frontend_flutter/lib/models/course.dart
class Course {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final double overallProgress;
  final double? averageGrade;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Course({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.overallProgress,
    this.averageGrade,
    required this.createdAt,
    this.updatedAt,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'],
      userId: json['user_id'],
      title: json['title'],
      description: json['description'],
      overallProgress: (json['overall_progress'] ?? 0.0).toDouble(),
      averageGrade: json['average_grade']?.toDouble(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'title': title,
      'description': description,
      'overall_progress': overallProgress,
      'average_grade': averageGrade,
      'created_at': createdAt.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
    };
  }
}
